File: mazeman/GAOverBuilds/Huffman96.py
#code message from byte array to printable symbol string
# in constructor set alphabet
# for compress use Huffman algorithm
# zero step - calculate number chars of alphabete inside data
# first step write lengths of dictionary's element
# second step write dictionary
#dictionary in format x^n[0]+(1-x)^n[1]...![+...] where n = vector of alphabet elements length, a = vector of codes each alphabet element, x = a[0].length % 2, n[0] != a[0].length => n[0] = (a[0].length + 1) / 2, a[0].length = n[0] * 2 - 1 ; n[i] = a[i].length, i >= 1
# third step - length of message in 15th base
#4th digit sequence transform to int ___ if next item is 1111 then it is end of sequence
# 4th step - code all data in message
import io
from collections import deque
class Huffman96(object):
    incomeAlphabet = None
    alphabetLong = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_";
    alphabetShort = "`abcdefghijklmnopqrstuvwxyz{|}~\t";

    def __init__(self, iA):
        t0 = str(type(iA))
        self.incomeAlphabet = []
        if t0 == "<class 'NoneType'>":
            for i0 in range(256):
                self.incomeAlphabet.append(i0)
        elif t0 == "<class 'str'>":
            for i0 in iA:
                self.incomeAlphabet.append(ord(i0))
        elif t0 == "<class 'list'>":
            incomeAlphabet = iA.copy()
        elif t0 == "<class 'bytes'>":
            for i0 in iA:
                self.incomeAlphabet.append(int(i0))
        elif t0 == "<class 'bytearray'>":
            for i0 in iA:
                self.incomeAlphabet.append(int(i0))
        else:
            print("bad format of income alphabet")
            incomeAlphabet = None
        pass

    def Encode(self, msg):
        if type(msg) == type(""):
            msg = bytearray(msg, encoding = 'utf8')
        else:
            msg = bytearray(msg)
        tree0 = []  #tree of Node fill all possible in order of incomeAlphabet
        for i0 in self.incomeAlphabet:
            tree0.append(Node(i0))
        #count all element from message
        for i0 in msg:
            tree0[self.incomeAlphabet.index(i0)].inc()
        #colapse tree
        while len(tree0) > 1:
            last2elem = []
            for i0 in tree0:
                if len(last2elem) == 0:
                    last2elem.append(i0)
                elif len(last2elem) == 1:
                    if last2elem[0].number > i0.number:
                        last2elem.append(last2elem[0])
                        last2elem[0] = i0
                    else:
                        last2elem.append(i0)
                else:
                    if i0.number < last2elem[0].number:
                        last2elem[1] = last2elem[0]
                        last2elem[0] = i0
                    elif i0.number < last2elem[1].number:
                        last2elem[1] = i0
            tree0.append(Node(-1).Insert(last2elem[1], last2elem[0]))
            tree0.remove(last2elem[1])
            tree0.remove(last2elem[0])
        #decolapse tree
        dict0 = tree0[0].Codize("", {})
        res0 = io.StringIO()
        buf0 = deque([], None)
        #code legend
        x = (len(dict0[self.incomeAlphabet[0]]) % 2) == 1
        n0 = (len(dict0[self.incomeAlphabet[0]]) + 1) // 2
        while n0 > 0:
            buf0.append(x)
            n0 -= 1
        for i0 in range(1, len(self.incomeAlphabet)):
            x = not x
            n0 = len(dict0[self.incomeAlphabet[i0]])
            while n0 > 0:
                buf0.append(x)
                n0 -= 1
        x = not x
        buf0.append(x)
        for i0 in range(len(self.incomeAlphabet)):
            for i1 in dict0[self.incomeAlphabet[i0]]:
                buf0.append(i1 == "1")
        #code length of message
        x = len(msg)
        buf1 = deque([True, True, True, True], None)
        while x > 0:
            j0 = x % 15
            x //= 15
            for i0 in range(4):
                buf1.appendleft(j0 % 2 == 1)
                j0 //= 2
        buf0.extend(buf1)
        #same encode message
        for i0 in msg:
            for i1 in dict0[i0]:
                buf0.append(i1 == "1")
            while len(buf0) > 6:
                if buf0.popleft():
                    n0 = 6
                    j0 = 0
                    while n0 > 0:
                        j0 *= 2
                        if buf0.popleft():
                            j0 += 1
                        n0 -= 1
                    res0.write(self.alphabetLong[j0])
                else:
                    n0 = 5
                    j0 = 0
                    while n0 > 0:
                        j0 *= 2
                        if buf0.popleft():
                            j0 += 1
                        n0 -= 1
                    res0.write(self.alphabetShort[j0])
        buf0.append(True)
        for i0 in range(10):
            buf0.append(False)
        while len(buf0) > 6:
            if buf0.popleft():
                n0 = 6
                j0 = 0
                while n0 > 0:
                    j0 *= 2
                    if buf0.popleft():
                        j0 += 1
                    n0 -= 1
                res0.write(self.alphabetLong[j0])
            else:
                n0 = 5
                j0 = 0
                while n0 > 0:
                    j0 *= 2
                    if buf0.popleft():
                        j0 += 1
                    n0 -= 1
                res0.write(self.alphabetShort[j0])
        return res0.getvalue()

    def Binarize(self, smb):
        if smb in self.alphabetLong:
            n0 = self.alphabetLong.index(smb)
            res0 = ""
            n1 = 6
            while n1 > 0:
                res0 = ("1" if n0 % 2 == 1 else "0") + res0
                n0 //= 2
                n1 -= 1
            return "1" + res0
        else:
            n0 = self.alphabetShort.index(smb)
            res0 = ""
            n1 = 5
            while n1 > 0:
                res0 = ("1" if n0 % 2 == 1 else "0") + res0
                n0 //= 2
                n1 -= 1
            return "0" + res0

    def Decode(self, msg):
        j0 = 0
        buf0 = self.Binarize(msg[j0])
        j0 += 1
        j1 = 0 #index of begin seqence
        j2 = 1 #index of end sequence
        #detect codes lengths
        x = buf0[0:1] == "1"
        lng0 = []
        for i0 in range(len(self.incomeAlphabet)):
            while buf0[j1] == buf0[j2]:
                j2 += 1
                while j2 + 1 >= len(buf0):
                    buf0 += self.Binarize(msg[j0])
                    j0 += 1
            j3 = j2 - j1
            if i0 == 0:
                lng0.append(j3 * 2 - 1 + (0 if x else 1))
            else:
                lng0.append(j3)
            j1 = j2
            j2 = j1 + 1
        while j2 + 1 >= len(buf0):
            buf0 += self.Binarize(msg[j0])
            j0 += 1
        buf0 = buf0[j2:]
        #fill dictionary of codes & find max length
        j1 = 0  #max
        dict0 = {}
        for i0 in range(len(lng0)):
            while len(buf0) <= lng0[i0]:
                buf0 += self.Binarize(msg[j0])
                j0 += 1
            dict0[buf0[0:lng0[i0]]] = chr(self.incomeAlphabet[i0])
            buf0 = buf0[lng0[i0]:]
            if lng0[i0] > j1:
                j1 = lng0[i0]
        # detect message length
        j3 = 0
        while True:
            while len(buf0) <= 4:
                buf0 += self.Binarize(msg[j0])
                j0 += 1
            if buf0[0:4] == "1111":
                buf0 = buf0[4:]
                break
            else:
                j3 *= 15
                j2 = 0
                for i0 in range(4):
                    j2 *= 2
                    j2 += (1 if buf0[i0:i0 + 1] == "1" else 0)
                buf0 = buf0[4:]
                j3 += j2
        # same message decode
        res0 = io.StringIO()
        while j0 < len(msg) and j3 > 0:
            while len(buf0) <= j1:
                buf0 += self.Binarize(msg[j0])
                j0 += 1
            j2 = j1
            while j2 > 0:
                if buf0[0:j2] in dict0:
                    res0.write(dict0[buf0[0:j2]])
                    buf0 = buf0[j2:]
                    j3 -= 1
                    break
                j2 -= 1
            if j2 == 0:
                break
        return res0.getvalue()

class Node(object):
    def __init__(self, symb):
        self.upper = None   #upper branch get append 1 on collapse tree
        self.lower = None   #lower branch get append 0 on collapse tree
        self.symbol = int(symb) #same symbol
        self.code = ""      #string of 0 and 1 's that put in output instead his symbol
        self.number = 0     #count how much his symbol in message

    def inc(self):
        self.number += 1
        return None

    def Codize(self, code, dict):
        self.code = code
        if self.upper == None and self.lower == None:
            dict[self.symbol] = self.code
        else:
            dict = self.upper.Codize(self.code + "1", dict)
            dict = self.lower.Codize(self.code + "0", dict)
        return dict

    def Insert(self, up, down):
        self.upper = up
        self.lower = down
        self.number = up.number + down.number
        return self

File: mazeman/GAOverBuilds/Huffman96.js
/**
 *code message from byte array to printable symbol string
 * in constructor set alphabet
 * for compress use Huffman algorithm
 * zero step - calculate number chars of alphabete inside data
 * first step write lengths of dictionary's element
 * second step write dictionary
 *dictionary in format x^n[0]+(1-x)^n[1]...![+...] where n = vector of alphabet elements length, a = vector of codes each alphabet element, x = a[0].length % 2, n[0] != a[0].length => n[0] = (a[0].length + x) / 2, a[0].length = n[0] * 2 - x ; n[i] = a[i].length, i >= 1
 * third step - length of message in 15th base
 *4th digit sequence transform to int ___ if next item is 1111 then it is end of sequence
 * 4th step - code all data in message
 **/
window.Huffman96 = function(a) {
  this.incomeAlphabet = undefined;
  this.alphabetLong = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_";
  this.alphabetShort = "`abcdefghijklmnopqrstuvwxyz{|}~\t";
  switch (typeof a) {
    case "undefined":
      this.incomeAlphabet = [];
      for (var i0 = 0; i0 < 256; i0++) this.incomeAlphabet.push(i0);
      break;
    case "string":
      this.incomeAlphabet = [];
      for (var i0 = 0; i0 < a.length; i0++)
        if (!this.incomeAlphabet.includes(a.charCodeAt(i0)))
          this.incomeAlphabet.push(a.charCodeAt(i0));
      break;
    case "object":
      if (a.__proto__.constructor.name == "Uint8Array") {
        this.incomeAlphabet = [];
        for (var i0 = 0; i0 < a.length; i0++)
          if (!this.incomeAlphabet.includes(a[i0]))
            this.incomeAlphabet.push(a[i0]);
      }
      if (a.__proto__.constructor.name == "Array") {
        this.incomeAlphabet = a;
      }
      //console.log(a.__proto__.constructor.name);
      break;
    default:
      console.log(typeof a);
      break;
  }

  this.CNode = function(smb) {
    this.upper = undefined;
    this.lower = undefined;
    this.symb = smb;
    this.code = "";
    this.count = 0;

    this.Inc = function() {
      this.count++;
    };

    this.Codize = function(c, d) {
      this.code = c;
      if (this.upper == undefined && this.lower == undefined) {
        d[this.symb] = this.code;
      } else {
        d = this.upper.Codize(this.code + "1", d);
        d = this.lower.Codize(this.code + "0", d);
      }
      return d;
    };

    this.Insert = function(u, l) {
      this.upper = u;
      this.lower = l;
      this.count = u.count + l.count;
      return this;
    };
  };

  this.Encode = function(msg) {
    //count of each content element
    var tree0 = {};
    for (var i0 = 0; i0 < this.incomeAlphabet.length; i0++) tree0[i0] = new this.CNode(this.incomeAlphabet[i0]);
    switch (msg.__proto__.constructor.name) {
      case "String":
        var msgt = [];
        for (var i0 = 0; i0 < msg.length; i0++) {
          var j0 = this.incomeAlphabet.indexOf(msg.charCodeAt(i0));
          msgt.push(msg.charCodeAt(i0));
          tree0[j0].Inc();
        }
        msg = msgt;
        break;
      case "Array":
        var j1 = 0;
        var msgt = [];
        for (var i0 = 0; i0 < msg.length; i0++) {
          var j0 = this.incomeAlphabet.indexOf(msg[i0]);
          if (j0 < 0) {
            msg[i0] = undefined;
            if (i0 - 1 > j1) {
              msgt = msgt.concat(msg.slice(j1, i0));
            }
            j1 = i0 + 1;
          } else {
            tree0[j0].Inc();
          }
        }
        if (j1 > 0) {
          if (msg.length - 1 > j1) {
            msgt = msgt.concat(msg.slice(j1, msg.length));
          }
          msg = msgt;
        }
        break;
      default:
        console.log("Incorrect type of income message!");
        return "";
    }
    //collapse
    var keys0 = Object.keys(tree0);
    var j0 = -1;
    while (keys0.length > 1 && keys0.length < 260) {
      var last2elem = [];
      for (var i0 = 0; i0 < keys0.length; i0++) {
        if (last2elem.length == 0) {
          last2elem.push(tree0[keys0[i0]]);
        } else {
          if (last2elem.length == 1) {
            if (last2elem[0].count > tree0[keys0[i0]].count) {
              last2elem.unshift(tree0[keys0[i0]]);
            } else {
              last2elem.push(tree0[keys0[i0]]);
            }
          } else {
            if (tree0[keys0[i0]].count < last2elem[1].count) {
              if (tree0[keys0[i0]].count < last2elem[0].count) {
                last2elem.pop();
                last2elem.unshift(tree0[keys0[i0]]);
              } else {
                last2elem.pop();
                last2elem.push(tree0[keys0[i0]]);
              }
            }
          }
        }
      }
      tree0[j0] = (new this.CNode(j0)).Insert(last2elem[1], last2elem[0]);
      j0--;
      var j1 = this.incomeAlphabet.indexOf(last2elem[0].symb);
      if (j1 > -1) delete tree0[j1];
      else delete tree0[last2elem[0].symb];
      j1 = this.incomeAlphabet.indexOf(last2elem[1].symb);
      if (j1 > -1) delete tree0[j1];
      else delete tree0[last2elem[1].symb];
      keys0 = Object.keys(tree0);
    }
    //decollapse tree
    var dict0 = tree0[keys0[0]].Codize("", {});
    tree0 = null;
    var buf0 = "";
    //write alphabet
    var x = dict0[this.incomeAlphabet[0]].length % 2;
    j0 = Math.floor((dict0[this.incomeAlphabet[0]].length + x) / 2);
    for (var i0 = 0; i0 < j0; i0++) buf0 += x.toString();
    x = 1 - x;
    for (var i0 = 1; i0 < this.incomeAlphabet.length; i0++) {
      j0 = dict0[this.incomeAlphabet[i0]].length;
      for (var i1 = 0; i1 < j0; i1++) buf0 += x.toString();
      x = 1 - x;
    }
    buf0 += x.toString();
    for (var i0 = 0; i0 < this.incomeAlphabet.length; i0++) {
      buf0 += dict0[this.incomeAlphabet[i0]];
    }
    //write msg length
    x = "1111";
    j0 = msg.length;
    while (j0 > 0) {
      var j1 = j0 % 15;
      for (var i0 = 0; i0 < 4; i0++) {
        x = (j1 % 2).toString() + x;
        j1 = Math.floor(j1 / 2);
      }
      j0 = Math.floor(j0 / 15);
    }
    buf0 += x;
    //same msg coding
    var res0 = {
      out: "",
      buf: buf0
    };
    for (var i0 = 0; i0 < msg.length; i0++) {
      res0.buf += dict0[msg[i0]];
      res0 = this.Unbinarize(res0);
    }
    while (res0.buf.length < 16) res0.buf += "0";
    res0 = this.Unbinarize(res0);
    return res0.out;
  };

  this.Decode = function(msg) {
    if (typeof msg != "string") {
      console.log("incorrect input data!");
      return [];
    }
    var buf0 = this.Binarize(msg.charAt(0));
    var j0 = 1; // index of message
    var j1 = 0; // index of begin sequence
    var j2 = 1; // index of end sequence
    //detect codes length
    var x = buf0.startsWith("1");
    var lng0 = [];
    var j3 = -1;
    for (var i0 = 0; i0 < this.incomeAlphabet.length; i0++) {
      while (buf0[j1] == buf0[j2]) {
        j2++;
        while (j2 + 1 >= buf0.length) {
          buf0 += this.Binarize(msg.charAt(j0));
          j0++;
        }
      }
      j3 = j2 - j1;
      if (i0 == 0) {
        lng0.push(j3 * 2 - ((x) ? 1 : 0));
      } else {
        lng0.push(j3);
      }
      j1 = j2;
      j2++;
    }
    while (j2 + 1 >= buf0.length) {
      buf0 += this.Binarize(msg.charAt(j0));
      j0++;
    }
    buf0 = buf0.substring(j2);
    //fill dictionary by lengths & find max length
    j1 = 0; //max
    var dict0 = {};
    for (var i0 = 0; i0 < lng0.length; i0++) {
      while (buf0.length <= lng0[i0]) {
        buf0 += this.Binarize(msg.charAt(j0));
        j0++;
      }
      dict0[buf0.substring(0, lng0[i0])] = this.incomeAlphabet[i0];
      buf0 = buf0.substring(lng0[i0]);
      if (lng0[i0] > j1) {
        j1 = lng0[i0];
      }
    }
    //detect message length
    j3 = 0; //leength value here
    while (true) {
      while (buf0.length <= 4) {
        buf0 += this.Binarize(msg.charAt(j0));
        j0++;
      }
      if (buf0.substring(0, 4) == "1111") {
        buf0 = buf0.substring(4);
        break;
      } else {
        j3 *= 15;
        j2 = 0;
        for (var i0 = 0; i0 < 4; i0++) {
          j2 *= 2;
          if (buf0.charAt(i0) == "1") j2++;
        }
        buf0 = buf0.substring(4);
        j3 += j2;
      }
    }
    //same message decode
    var res0 = [];
    while (j3 > 0) {
      while (buf0.length < j1) {
        buf0 += this.Binarize(msg.charAt(j0));
        j0++;
      }
      j2 = j1;
      while (j2 > 0) {
        if (dict0.hasOwnProperty(buf0.substring(0, j2))) {
          res0.push(dict0[buf0.substring(0, j2)]);
          buf0 = buf0.substring(j2);
          j3--;
          break;
        }
        j2--;
        if (j2 == 0) {
          console.log("Some error on decode");
          break;
        }
      }
    }
    return res0;
  };

  this.Unbinarize = function(x) {
    while (x.buf.length > 7) {
      var j0 = 0;
      if (x.buf.startsWith("1")) {
        for (var i0 = 1; i0 < 7; i0++) {
          j0 *= 2;
          if (x.buf.charAt(i0) == "1") j0++;
        }
        x.out += this.alphabetLong.charAt(j0);
        x.buf = x.buf.substring(7);
      } else { //"0"
        for (var i0 = 1; i0 < 6; i0++) {
          j0 *= 2;
          if (x.buf.charAt(i0) == "1") j0++;
        }
        x.out += this.alphabetShort.charAt(j0);
        x.buf = x.buf.substring(6);
      }
    }
    return x;
  };

  this.Binarize = function(x) {
    if (this.alphabetLong.includes(x)) {
      var j0 = this.alphabetLong.indexOf(x);
      var s0 = "";
      for (var i0 = 0; i0 < 6; i0++) {
        s0 = (j0 % 2).toString() + s0;
        j0 = Math.floor(j0 / 2);
      }
      return "1" + s0;
    }
    if (this.alphabetShort.includes(x)) {
      var j0 = this.alphabetShort.indexOf(x);
      var s0 = "";
      for (var i0 = 0; i0 < 5; i0++) {
        s0 = (j0 % 2).toString() + s0;
        j0 = Math.floor(j0 / 2);
      }
      return "0" + s0;
    }
  };
};
